src/main/java/dev/harrel/jsonschema/ValidatorFactory.java

```diff
@@ -53,14 +55,33 @@ public ValidatorFactory withEvaluatorFactory(EvaluatorFactory evaluatorFactory)
     }
 
     /**
-     * Sets {@link JsonNodeFactory}. Provided default is {@link JacksonNode.Factory}.
+     * Sets {@link JsonNodeFactory} to be used for both schema and data parsing.
+     * If you require different factories for schemas and data, please use {@link ValidatorFactory#withJsonNodeFactories}
+     * Provided default is {@link JacksonNode.Factory}.
      *
      * @param jsonNodeFactory {@code JsonNodeFactory} to be used
      * @return self
      */
     public ValidatorFactory withJsonNodeFactory(JsonNodeFactory jsonNodeFactory) {
-        Objects.requireNonNull(jsonNodeFactory);
-        this.jsonNodeFactory = () -> jsonNodeFactory;
+        return withJsonNodeFactories(jsonNodeFactory, jsonNodeFactory);
+    }
+
+    /**
+     * Sets one {@link JsonNodeFactory} for schema parsing and one for data parsing.
+     * Might be useful when you expect schemas to be in a different format than data (JSON/YAML).
+     * In most cases having two different factories is not required,
+     * so please just use {@link ValidatorFactory#withJsonNodeFactory} whenever possible.
+     * Provided default is {@link JacksonNode.Factory}.
+     *
+     * @param schemaNodeFactory   {@code JsonNodeFactory} to be used for parsing schemas
+     * @param instanceNodeFactory {@code JsonNodeFactory} to be used for parsing data
+     * @return self
+     */
+    public ValidatorFactory withJsonNodeFactories(JsonNodeFactory schemaNodeFactory, JsonNodeFactory instanceNodeFactory) {
```

Comment:
@harrel56 Does this have to be coupled? If I have a `ValidatorFactory` instance and I want to change the schema factory, I have to know the instance factory too, which creates strange data flow and responsibility issues.

For example:
```java
void validate(v: ValidatorFactory) {
    v.withJsonNodeFactories(myJsonParsers(), ??? what do I write here ???)
    v.createValidator()...
}
```
or the other way around:
```java
void validate() {
     ValidatorFactory v = new ValidatorFactory()
          .withJsonNodeFactory(mySpecialParser())
          .createValidator()
```
where I would expect for it to "just work" with the schema parsing, but I've silently overridden the schema parser too. The default JSON one would've been good.

I understand you might be doing this to maintain compatibility, but I see you're cutting a 2.x anyway, so it's a good time to decouple these two and let the user set only the one that they want.

Reply:
Note, if you want to have the coupled API be the default, I would recommend these signatures:
```java
// Original method with changed behavior:
public ValidatorFactory withJsonNodeFactory(JsonNodeFactory jsonNodeFactory) {
    withInstanceNodeFactory(jsonNodeFactory);
    withSchemaNodeFactory(jsonNodeFactory);
}

// Decoupled new setters for the two separate conpects:
public ValidatorFactory withInstanceNodeFactory(JsonNodeFactory jsonNodeFactory) {
    this.instanceNodeFactory = () -> jsonNodeFactory;
}
public ValidatorFactory withSchemaNodeFactory(JsonNodeFactory jsonNodeFactory) {
    this.schemaNodeFactory = () -> jsonNodeFactory;
}
```

Reply:
Thanks for your feedback! TBH I didn't consider this coupling to be a bigger issue. From my point of view, if someone wants to manage schema/instance factories separately they are kind of required to know what are they doing and what the second factory should be. I know this doesn't necessarily work like that in a code but well.

And as for the suggestion from your second comment: this is definitely a viable approach. Yes, it's, as you say, "decoupled" but I was worried about one thing. If someone wants to parse only data as YAML he would write:
```java
     ValidatorFactory v = new ValidatorFactory()
          .withInstanceNodeFactory(mySpecialParser())
          .createValidator()
```
This means he wants to have a default schema factory (which is jackson), but what are the chances he doesn't really know that? And if there is no jackson lib in classpath this code would just fail in runtime. So if anyone is changing only one factory, there is always risk involved. Hence I guess the approach I took - this way if anyone wants to have different factories he needs to have knowledge about both of them and consciously pick the ones which are on classpath.

Also, these 2 signatures are pretty similar `withJsonNodeFactory` vs `withInstanceNodeFactory`. Accidentally using the latter would cause the same problems and rather unexpected behavior. In the current approach such mistake is not possible.

Still, gonna give it more thought before releasing. Maybe there is yet another solution :)

Reply:
@TWiStErRob FYI decided to leave it as it is for now. Mainly because of the risk of accidental misconfiguration I mentioned above. Nevertheless, I appreciate your input

Reply:
Taking a step back, could we maybe have the two separate/independent methods for setting them **in addition** to the coupled one(s)?

My use case being: I do not want my custom yaml parser to be used for JSON schema. i.e. I do not want to override the default, don't even want to know what it is, I trust you set it up nicely. My only way at the moment is to call the two-parameter version, but what do I write in the schema parser param? Is there a default factory exposed somewhere?

Reply:
> I do not want my custom yaml parser to be used for JSON schema. i.e. I do not want to override the default, don't even want to know what it is, I trust you set it up nicely.

Well, this is the whole point behind it - you have to worry about what to put in here so you won't get `NoClassDefFoundError` in runtime. The default JSON provider is [jackson](https://mvnrepository.com/artifact/com.fasterxml.jackson.core/jackson-databind) - do you have it on your classpath? You can also check out this section https://github.com/harrel56/json-schema?tab=readme-ov-file#jsonyaml-providers, for each provider there is a factory class listed. 

Just to clarify (maybe there's a misunderstanding): you have to have an explicit 3rd party dependency to JSON (or YAML now) dependency in your project in order for this lib to work correctly. Jackson being default is just because I thought it'll be most common and I wanted to spare most users this 1 additional configuration line.

Reply:
Looking at your code I'm assuming you'd like to use `kotlinx-serialization-json` probably. So, there's no factory for that unfortunately, you'd have to write your own implementation as you did for snakeyaml engine. NGL adding official support for kotlin JSON lib is tempting, I was considering it at some point. Probably should've supported this rather than some strange jettison libs :)